python/src/nanoarrow/c_array.py

```diff
@@ -0,0 +1,573 @@
+
+    This builder converts an iterable to a CArray using some heuristics to pick
+    the fastest available method for converting to a particular type of array.
+    Briefly, the methods are (1) ArrowArrayAppendXXX() functions from the C
+    library (string, binary), (2) array.array() (integer/float except float16),
+    (3) CBufferBuilder.write_elements() (everything else).
+    """
+
+    @classmethod
+    def infer_schema(cls, obj) -> Tuple[CBuffer, CSchema]:
+        raise ValueError("schema is required to build array from iterable")
+
+    def __init__(self, schema):
+        super().__init__(schema)
+
+        # Resolve the method name we are going to use to do the building from
+        # the provided schema.
+        type_id = self._schema_view.type_id
+        if type_id not in _ARRAY_BUILDER_FROM_ITERABLE_METHOD:
+            raise ValueError(
+                f"Can't build array of type {self._schema_view.type} from iterable"
+            )
+
+        method_name = _ARRAY_BUILDER_FROM_ITERABLE_METHOD[type_id]
+
+        # If there might be nulls, we may need to pick a different strategy
+        if (
+            self._schema_view.nullable
+            and method_name in _ARRAY_BUILDER_FROM_NULLABLE_ITERABLE_METHOD
+        ):
+            method_name = _ARRAY_BUILDER_FROM_NULLABLE_ITERABLE_METHOD[method_name]
+
+        self._append_impl = getattr(self, method_name)
+
+    def start_building(self, c_builder: CArrayBuilder) -> None:
+        c_builder.start_appending()
+
+    def append(self, c_builder: CArrayBuilder, obj: Any) -> None:
+        self._append_impl(c_builder, obj)
+
+    def _append_strings(self, c_builder: CArrayBuilder, obj: Iterable) -> None:
+        c_builder.append_strings(obj)
+
+    def _append_bytes(self, c_builder: CArrayBuilder, obj: Iterable) -> None:
+        c_builder.append_bytes(obj)
+
+    def _build_nullable_array_using_array(
+        self, c_builder: CArrayBuilder, obj: Iterable
+    ) -> None:
+        wrapper = NoneAwareWrapperIterator(
+            obj, self._schema_view.storage_type_id, self._schema_view.fixed_size
+        )
+        self._append_using_array(c_builder, wrapper)
+
+        _, null_count, validity = wrapper.finish()
+        if validity is not None:
+            c_builder.set_buffer(0, validity, move=True)
+
+        c_builder.set_null_count(null_count)
+
+    def _build_nullable_array_using_buffer_builder(
+        self, c_builder: CArrayBuilder, obj: Iterable
+    ) -> None:
+        wrapper = NoneAwareWrapperIterator(
+            obj, self._schema_view.storage_type_id, self._schema_view.fixed_size
+        )
+        self._append_using_buffer_builder(c_builder, wrapper)
+
+        _, null_count, validity = wrapper.finish()
+        if validity is not None:
+            c_builder.set_buffer(0, validity, move=True)
+
+        c_builder.set_null_count(null_count)
+
+    def _append_using_array(self, c_builder: CArrayBuilder, obj: Iterable) -> None:
```

Comment:
What's the difference between using array vs using buffer builder? The latter is for the ones there is no matching buffer protocol type?